google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddDynamicPageFeed.java

```diff
@@ -0,0 +1,543 @@
+        FeedAttribute.newBuilder()
+            .setType(FeedAttributeType.URL_LIST)
+            .setName(StringValue.of("Page URL"))
+            .build();
+
+    // Create a label attribute
+    FeedAttribute labelAttribute =
+        FeedAttribute.newBuilder()
+            .setType(FeedAttributeType.STRING_LIST)
+            .setName(StringValue.of("Label"))
+            .build();
+
+    // Create the feed.
+    Feed feed =
+        Feed.newBuilder()
+            .setName(StringValue.of("DSA Feed #" + System.currentTimeMillis()))
+            .addAttributes(urlAttribute)
+            .addAttributes(labelAttribute)
+            .setOrigin(FeedOrigin.USER)
+            .build();
+
+    // Create the operation.
+    FeedOperation operation = FeedOperation.newBuilder().setCreate(feed).build();
+
+    // Create the feed service client.
+    try (FeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createFeedServiceClient()) {
+      // Add the feed.
+      MutateFeedsResponse response =
+          feedServiceClient.mutateFeeds(String.valueOf(customerId), ImmutableList.of(operation));
+      String feedResourceName = response.getResults(0).getResourceName();
+      // Display the result.
+      System.out.printf("Feed with resource name '%s' was created.%n", feedResourceName);
+
+      // Return a FeedDetails object.
+      return getFeed(googleAdsClient, customerId, feedResourceName);
+    }
+  }
+
+  /**
+   * Retrieves details about a feed.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID in which to create criterion.
+   * @return a FeedDetails object containing relevant information.
+   */
+  private FeedDetails getFeed(
+      GoogleAdsClient googleAdsClient, long customerId, String feedResourceName) {
+    // Construct the query.
+    String query =
+        "SELECT feed.attributes FROM feed WHERE feed.resource_name = '" + feedResourceName + "'";
+
+    // Construct the request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(String.valueOf(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .build();
+
+    // Issue the search request.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      SearchPagedResponse searchPagedResponse = googleAdsServiceClient.search(request);
+      // Get the first result because we only need the single feed item we created previously.
+      GoogleAdsRow googleAdsRow = searchPagedResponse.getPage().getResponse().getResults(0);
+      // Get the attributes list from the feed and create a map with keys of each attribute and
+      // values of each corresponding ID.
+      List<FeedAttribute> feedAttributeList = googleAdsRow.getFeed().getAttributesList();
+      Map<String, Long> feedAttributeMap = new Hashtable<>();
```

Comment:
Please use HashMap instead. Hashtable is left over from the very early days of Java and is rarely used these days.

Reply:
Done.